### ISSUE TITLE
adding dependencies sometimes breaks completions

### ISSUE THREAD

**Issue by johnw42** [rust-analyzer/rust-analyzer#9584]
I've been using rust-analyzer through VSCode and I keep encountering situations where adding a dependency causes some completions to stop working. I've attached the smallest example I could find, where completing the word "info" in main.rs should suggest `log::info`, and it does if either `x11` or `clap` is removed from the dependencies, but with both dependencies in place, the only suggestions are `std::panic::PanicInfo` and a bunch of types from `x11` whose names end in `Info`.

I'm using the following versions:

VSCode: 1.58.0 2d23c42a936db1c7b3b06f918cde29561cc47cd6
rust-analyzer extension: 0.2.662
rust-analyzer e5c1c8cf2 2021-07-05 stable
cargo 1.52.0 (69767412a 2021-04-21)
rustc 1.52.1 (9bc8c42bb 2021-05-09)

[example.tar.gz](https://github.com/rust-analyzer/rust-analyzer.github.io/files/6794436/example.tar.gz)

**Comment by jonas-schievink** [comment]
There's a limit to how many results we process, you might be hitting that

**Comment by flodiebold** [comment]
We should try to rank results better before limiting them (prefer exact case matches, prefer matches in the beginning of the name).

**Comment by johnw42** [comment]
Yeah, that's more or less what I was going to suggest. I was surprised to
look in the code and see that it seems to completely ignore the text before
the cursor when building its list of completions.

**Comment by flodiebold** [comment]
It doesn't completely ignore the text before the cursor, but it does cut off the number of results before ranking them.

**Comment by mahdi-frms** [comment]
@flodiebold could you provide some instructions to fix this?
